File: src-react/src/lib/api/projects.ts
import type { GitBaseFolder, Project, ScanResult } from "@/types";
import { invokeCommand } from "@/lib/api/tauri";

export const projectsApi = {
  list(): Promise<Project[]> {
    return invokeCommand<Project[]>("list_projects");
  },

  record(path: string): Promise<Project> {
    return invokeCommand<Project>("record_project", { path });
  },

  delete(id: string): Promise<void> {
    return invokeCommand<void>("delete_project", { id });
  },

  reorder(projectIds: string[]): Promise<Project[]> {
    return invokeCommand<Project[]>("reorder_projects", { projectIds });
  },

  listGitBaseFolders(): Promise<GitBaseFolder[]> {
    return invokeCommand<GitBaseFolder[]>("list_git_base_folders");
  },

  recordGitBaseFolder(path: string): Promise<GitBaseFolder> {
    return invokeCommand<GitBaseFolder>("record_git_base_folder", { path });
  },

  removeGitBaseFolder(id: string): Promise<void> {
    return invokeCommand<void>("remove_git_base_folder", { id });
  },

  scanBaseFolder(path: string): Promise<ScanResult[]> {
    return invokeCommand<ScanResult[]>("scan_git_base_folder", { path });
  },

  scanGitBaseFolders(): Promise<ScanResult[]> {
    return invokeCommand<ScanResult[]>("scan_git_base_folders");
  },

  /** Replace the full set of Project custom skills directories. The backend
   *  normalizes and de-duplicates the list and rejects dirs that resolve to a
   *  fixed Agent project skills dir. Returns the updated Project. */
  setCustomSkillsDirs(projectId: string, dirs: string[]): Promise<Project> {
    return invokeCommand<Project>("set_project_custom_skills_dirs", { projectId, dirs });
  },

  /** Replace the full set of Project extra prompt files. The backend normalizes
   *  and de-duplicates the list and rejects entries whose filename does not match
   *  an Agent `projectPromptFile` glob (AGENTS*.md / CLAUDE*.md). Returns the
   *  updated Project. */
  setExtraPromptFiles(projectId: string, files: string[]): Promise<Project> {
    return invokeCommand<Project>("set_project_extra_prompt_files", { projectId, files });
  },

  /** Override the Project Session Directory. An empty string restores the default
   *  `__sessions` template. Session Directory stays single-valued by design.
   *  Returns the updated Project. */
  setSessionsDir(projectId: string, dir: string): Promise<Project> {
    return invokeCommand<Project>("set_project_sessions_dir", { projectId, dir });
  },
};

File: src-tauri/src/commands/projects.rs
use tauri::State;

use nexus_core::{
    error::AppResult,
    services::projects::{DiscoveredRepo, GitBaseFolder, Project, ProjectDefaults},
};

use crate::store::AppState;

#[tauri::command]
pub fn list_projects(state: State<'_, AppState>) -> AppResult<Vec<Project>> {
    state.projects.list_projects()
}

#[tauri::command]
pub fn record_project(state: State<'_, AppState>, path: String) -> AppResult<Project> {
    state.projects.record_project(path)
}

#[tauri::command]
pub fn delete_project(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.projects.delete_project(id)
}

#[tauri::command]
pub fn reorder_projects(
    state: State<'_, AppState>,
    project_ids: Vec<String>,
) -> AppResult<Vec<Project>> {
    state.projects.reorder_projects(project_ids)
}

#[tauri::command]
pub fn list_git_base_folders(state: State<'_, AppState>) -> AppResult<Vec<GitBaseFolder>> {
    state.projects.list_git_base_folders()
}

#[tauri::command]
pub fn record_git_base_folder(
    state: State<'_, AppState>,
    path: String,
) -> AppResult<GitBaseFolder> {
    state.projects.record_git_base_folder(path)
}

#[tauri::command]
pub fn remove_git_base_folder(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.projects.remove_git_base_folder(id)
}

#[tauri::command]
pub fn scan_git_base_folder(
    state: State<'_, AppState>,
    path: String,
) -> AppResult<Vec<DiscoveredRepo>> {
    state.projects.scan_git_base_folder(path)
}

#[tauri::command]
pub fn scan_git_base_folders(state: State<'_, AppState>) -> AppResult<Vec<DiscoveredRepo>> {
    state.projects.scan_git_base_folders()
}

#[tauri::command]
pub fn set_project_custom_skills_dirs(
    state: State<'_, AppState>,
    project_id: String,
    dirs: Vec<String>,
) -> AppResult<Project> {
    state
        .projects
        .set_project_custom_skills_dirs(project_id, dirs)
}

#[tauri::command]
pub fn set_project_extra_prompt_files(
    state: State<'_, AppState>,
    project_id: String,
    files: Vec<String>,
) -> AppResult<Project> {
    state
        .projects
        .set_project_extra_prompt_files(project_id, files)
}

#[tauri::command]
pub fn set_project_sessions_dir(
    state: State<'_, AppState>,
    project_id: String,
    dir: String,
) -> AppResult<Project> {
    state.projects.set_project_sessions_dir(project_id, dir)
}

#[tauri::command]
pub fn get_project_defaults(state: State<'_, AppState>) -> AppResult<ProjectDefaults> {
    state.projects.get_project_defaults()
}

#[tauri::command]
pub fn set_default_custom_skills_dirs(
    state: State<'_, AppState>,
    dirs: Vec<String>,
) -> AppResult<ProjectDefaults> {
    state.projects.set_default_custom_skills_dirs(dirs)
}

#[tauri::command]
pub fn set_default_extra_prompt_files(
    state: State<'_, AppState>,
    files: Vec<String>,
) -> AppResult<ProjectDefaults> {
    state.projects.set_default_extra_prompt_files(files)
}

#[tauri::command]
pub fn set_default_sessions_dir(
    state: State<'_, AppState>,
    dir: String,
) -> AppResult<ProjectDefaults> {
    state.projects.set_default_sessions_dir(dir)
}
